### PR TITLE
Faster ONNX inference with bindings

### LINKED ISSUE #17127: ONNX is slower than Pytorch
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, 
I did benchmark COCO through the command below using pre-trained model.

`yolo benchmark model=/ultralytics/DATA/weights/coco_pretrained/yolov8/yolov8n.pt data=/ultralytics/DATA/weights/coco_pretrained/coco.yaml   
`      

The result is as shown in the image below.   
image

I think the result is a little strange here.
I know that onnx is usually faster than pytorch, but the benchmark result is the opposite.   
   
   
I'm using ultralytics through docker image(https://hub.docker.com/layers/ultralytics/ultralytics/latest/images/sha256-2d498f6677bcc7ccaf5ebda1c3091e1c683e10efe82591be47b3ec1904c86f02?context=explore).

`docker pull ultralytics/ultralytics:latest`   

So, my ultralytics version is "8.3.18"
   

May I know why onnx is slower than pytorch in your docker image environment?

`yolo checks`
Ultralytics 8.3.18 🚀 Python-3.11.9 torch-2.4.1+cu121 CUDA:0 (NVIDIA GeForce RTX 4090, 24217MiB)
Setup complete ✅ (32 CPUs, 125.5 GB RAM, 1549.8/1832.2 GB disk)

OS                  Linux-6.8.0-45-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.11.9
Install             git
RAM                 125.51 GB
Disk                1549.8/1832.2 GB
CPU                 Intel Core(TM) i9-14900K
CPU count           32
GPU                 NVIDIA GeForce RTX 4090, 24217MiB
GPU count           2
CUDA                12.1

numpy               ✅ 1.23.5>=1.23.0
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.2.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.4.1+cu121>=1.8.0
torchvision         ✅ 0.19.1+cu121>=0.9.0
tqdm                ✅ 4.66.5>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.9>=2.0.0
torch               ✅ 2.4.1+cu121!=2.4.0,>=1.8.0; sys_platform == "win32"

I installed onnx too.

image

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes https://github.com/ultralytics/ultralytics/issues/17127

Uses bindings to reduce CPU-GPU transfer overheads.

**Export**
```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")
model.export(format="onnx")
```

**Before changes**

```python
from ultralytics import YOLO

model = YOLO("yolo11n.onnx", task="detect")
results = model.val(data="coco.yaml", device=0)
```
```
Speed: 0.3ms preprocess, 9.0ms inference, 0.0ms loss, 1.6ms postprocess per image
```

**After**
```python
from ultralytics import YOLO

model = YOLO("yolo11n.onnx", task="detect")
results = model.val(data="coco.yaml", device=0)
```
```
Speed: 0.3ms preprocess, 5.9ms inference, 0.0ms loss, 1.6ms postprocess per image
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improved the ONNX Runtime integration to handle dynamic and static shapes more robustly with better CUDA support.

### 📊 Key Changes
- Updated provider logic to dynamically determine available ONNX Runtime execution providers.
- Implemented checks and warnings for CUDA execution fallbacks to CPU if CUDA is not available.
- Enhanced input-output binding for static shaped models to streamline execution with ONNX Runtime.
- Added bindings for output tensors directly to improve performance with specific configurations.

### 🎯 Purpose & Impact
- **Increased Flexibility:** The code now adapts to available hardware, allowing for more efficient model execution across different environments. 🔄
- **Improved Performance:** By leveraging optimized input-output bindings, model inference can be faster, especially for static models and configurations using CUDA. 🚀
- **User Alerts:** Clear warnings help users understand when their system might not support certain features, leading to more informed troubleshooting. ⚠️
- **Enhanced Usability:** Users with both dynamic and static models can expect better handling, reducing errors and improving overall experience. 🛠️

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help guide the review process, and an Ultralytics engineer will assist you soon.

To ensure a smooth integration of your contributions, please review the checklist below:

- ✅ **Define a Purpose**: Clearly articulate the purpose of your fix or enhancement in your PR description, and link to any relevant issues like [this one](https://github.com/ultralytics/ultralytics/issues/17127). Ensure your commit messages are precise and align with the project's standards.
- ✅ **Synchronize with Source**: Make sure your PR is up-to-date with the `main` branch of `ultralytics/ultralytics`. If there are discrepancies, you can resolve them by clicking the 'Update branch' button or by executing `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are successful. If any checks fail, please identify and fix the issues.
- ✅ **Update Documentation**: Update the [documentation](https://docs.ultralytics.com) as needed for any new or updated features.
- ✅ **Add Tests**: Include or modify tests to ensure your changes are adequately covered, and validate that all tests pass.
- ✅ **Sign the CLA**: If it's your inaugural PR with Ultralytics, please sign the [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) by stating, "I have read the CLA Document and I sign the CLA" in a new comment.
- ✅ **Minimize Changes**: Modify only the code essential for your bug fix or feature enhancement. 🧘‍♂️ As Bruce Lee said, _"It is not daily increase but daily decrease, hack away the unessential."_

For comprehensive guidance, take a look at our [Contributing Guide](https://docs.ultralytics.com/help/contributing). If you have any queries, don't hesitate to ask. Thank you for contributing to Ultralytics! 🚀✨

**Review by Burhan-Q — COMMENTED** [review]
Interesting! Looks like the benchmark tests are also seeing an improvement.

[This PR](https://github.com/ultralytics/ultralytics/actions/runs/11535563779/job/32110773088#step:6:121)

```
Speed: 9.8ms preprocess, 7.6ms inference, 8.8ms postprocess per image
```

[Another PR](https://github.com/ultralytics/ultralytics/actions/runs/11497963257/job/32002771867?pr=16955#step:6:120)

```
Speed: 4.7ms preprocess, 8.8ms inference, 0.6ms postprocess per image
```

That said, the preprocessing seems to be a bit higher, but that could be due to other overhead on that machine as your result doesn't appear to reflect the same. Ran a quick test with `yolo val model=yolo11n.(onnx|pt) data=coco128 device="cpu"`

```
# This PR ONNX (CPU)
Speed: 2.0ms preprocess, 219.6ms inference, 15.3ms postprocess per image

# On main PyTorch (CPU)
Speed: 1.7ms preprocess, 45.2ms inference, 1.7ms postprocess per image

# On main ONNX (CPU)
Speed: 1.8ms preprocess, 248.8ms inference, 4.3ms postprocess per image
```

Still doesn't beat PyTorch inference times, but does beat out the ONNX times, although in my quick test the post-processing is a bit slower.

**Comment by Y-T-G** [comment]
@Burhan-Q Interesting. I'm not able to get a consistent time on `coco128.yaml` on CPU. Sometimes the postprocessing is faster in the `main` and sometimes with this PR.

One of the runs:
`main`
`Speed: 1.1ms preprocess, 65.1ms inference, 0.0ms loss, 4.5ms postprocess per image`

`PR`
`Speed: 1.2ms preprocess, 61.1ms inference, 0.0ms loss, 3.0ms postprocess per image`

Particularly, if you load the model in Python and run val the second time, it's much better than the first time.

**Comment by Y-T-G** [comment]
@Laughing-q Hmm. It seems like your ONNX InferenceSession failed to use CUDAExexutionProvider.

https://github.com/ultralytics/ultralytics/blob/e1e54602dfbf6b9c52aaaa340681a5a8b8f089f4/ultralytics/nn/autobackend.py#L192-L193

Do you have onnxruntime-gpu?

Regardless, I should add a check and probably a warning too for this scenario.

**Comment by Laughing-q** [comment]
@Y-T-G ok I reinstalled the onnxruntime package and it works! I do get faster speed than main branch.

**Comment by Y-T-G** [comment]
@Laughing-q I added a check for the fallaback

**Comment by glenn-jocher** [comment]
@Y-T-G @Laughing-q maybe there's a better way to automatically handle the execution providers. I've updated it to now retrieve all available providers. On macOS for example there are additional ones here I didn't see before:
```
['CoreMLExecutionProvider', 'AzureExecutionProvider', 'CPUExecutionProvider']
```

**Comment by glenn-jocher** [comment]
@Y-T-G can you test this again after my changes? Thanks!

**Comment by glenn-jocher** [comment]
@Y-T-G @Laughing-q wow,  just independently allowing for a wider set of execution providers (just my last commit, not the rest of the PR) has this change (also slower pre and post processing but much faster inference)

Screenshot 2024-10-29 at 13 19 28

**Comment by Y-T-G** [comment]
@glenn-jocher I guess it's using CoreMLProvider in your case which should be faster than CPU

**Comment by glenn-jocher** [comment]
@Y-T-G PR merged!

**Comment by YOOHYOJEONG** [comment]
@Y-T-G 

Hi. 
I checked that the code is currently modified to this PR.
I tried benchmark again after pull with this version code, and I got the following error.


Loading /ultralytics/weights/coco_pretrained/yolov11/yolo11n.onnx for ONNX Runtime inference...
Preferring ONNX Runtime TensorrtExecutionProvider
2024-10-31 00:05:37.806269014 [E:onnxruntime:Default, provider_bridge_ort.cc:1978 TryGetProviderInfo_TensorRT] /onnxruntime_src/onnxruntime/core/session/provider_bridge_ort.cc:1637 onnxruntime::Provider& onnxruntime::ProviderLibrary::Get() [ONNXRuntimeError] : 1 : FAIL : Failed to load library libonnxruntime_providers_tensorrt.so with error: libnvinfer.so.10: cannot open shared object file: No such file or directory

*************** EP Error ***************
EP Error /onnxruntime_src/onnxruntime/python/onnxruntime_pybind_state.cc:490 void onnxruntime::python::RegisterTensorRTPluginsAsCustomOps(PySessionOptions&, const onnxruntime::ProviderOptions&) Please install TensorRT libraries as mentioned in the GPU requirements page, make sure they're in the PATH or LD_LIBRARY_PATH, and that your GPU is supported.
 when using ['TensorrtExecutionProvider', 'CUDAExecutionProvider', 'CPUExecutionProvider']
Falling back to ['CUDAExecutionProvider', 'CPUExecutionProvider'] and retrying.
****************************************   

May I know your onnx version and tensorRT version?


But even if this error occurs, it doesn't mean that it can't be inferred as onnx.
It appears to be deduced using GPU, much faster than the traditional main.

But it's still slower than pytorch..   
image

berfoe : 
https://github.com/ultralytics/ultralytics/issues/17127

**Comment by Y-T-G** [comment]
@YOOHYOJEONG The PyTorch model also runs with minimum rectangle, so it doesn't use the full 640x640 unlike ONNX which can give it an edge. You can try setting `rect=False` to `model.val` for PyTorch.

**Comment by YOOHYOJEONG** [comment]
@Y-T-G 
Hi. 

`rect=False` doesn't seem to progress to 640X640 even with this.
The actual output value is 384x640.

validate code   
'''python
from ultralytics import YOLO

# Load a model
model = YOLO("/ultralytics/DATA/weights/coco_pretrained/yolov11/yolo11n.pt")

# Customize validation settings
val_results = model.val(data="/ultralytics/coco.yaml", imgsz=640, rect=False, device="0")
'''
   
And if you look at the argument document(https://docs.ultralytics.com/modes/val/#arguments-for-yolo-model-validation), the default value is False, right?

**Comment by Y-T-G** [comment]
@YOOHYOJEONG You also probably need to change the `Letterbox` function to not use `auto=True` by default.

**Comment by YOOHYOJEONG** [comment]
@Y-T-G Oh.. I see. I'll try. Thank you for always helping me!
